### ISSUE TITLE
Analyzers should not register "async" callbacks with RegisterXXXAction 

### ISSUE THREAD

**Issue by mavasani** [icsharpcode/RefactoringEssentials#251]
Analyzer callbacks are generally invoked within an exception handler by the analyzer driver, so any exception from the analyzer is caught and reported as a diagnostic. However, if analyzer incorrectly registers an async method as a callback, any exception within the callback can crash VS (see https://github.com/code-cracker/code-cracker/issues/375#issue-87474837 for details).

The Roslyn analyzer framework team has had multiple reports of RefactoringEssentials analyzers crashing VS due to this issue. For example, FunctionNeverReturnsAnalyzer registers an async lambda callback [here](https://github.com/icsharpcode/RefactoringEssentials/blob/master/RefactoringEssentials/CSharp/Diagnostics/Synced/CodeQuality/FunctionNeverReturnsAnalyzer.cs#L36), and exception from the visitor crashes VS with below call stack:

```
Exception type:   System.NotSupportedException
Message:          Specified method is not supported.
InnerException:   <none>
StackTrace (generated):
    SP       IP       Function
    212DE610 2C3971DD RefactoringEssentials_25cb0000!RefactoringEssentials.Util.Analysis.ControlFlowGraphBuilder+NodeCreationVisitor.DefaultVisit(Microsoft.CodeAnalysis.SyntaxNode)+0x1d
    212DE618 2C3971AB Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor.VisitLocalFunctionStatement(Microsoft.CodeAnalysis.CSharp.Syntax.LocalFunctionStatementSyntax)+0xb
    212DE620 2C39433D Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.Syntax.LocalFunctionStatementSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor)+0xd
    212DE628 26251E54 Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor.Visit(Microsoft.CodeAnalysis.SyntaxNode)+0x2c
    212DE634 2625268C RefactoringEssentials_25cb0000!RefactoringEssentials.Util.Analysis.ControlFlowGraphBuilder+NodeCreationVisitor.HandleStatementList(System.Collections.Generic.IEnumerable`1<Microsoft.CodeAnalysis.CSharp.Syntax.StatementSyntax>, RefactoringEssentials.Util.Analysis.ControlFlowNode)+0xec
    212DE67C 26251F4F RefactoringEssentials_25cb0000!RefactoringEssentials.Util.Analysis.ControlFlowGraphBuilder+NodeCreationVisitor.VisitBlock(Microsoft.CodeAnalysis.CSharp.Syntax.BlockSyntax)+0x37
    212DE690 263BF8AD Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.Syntax.BlockSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor)+0xd
    212DE698 26251E54 Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor.Visit(Microsoft.CodeAnalysis.SyntaxNode)+0x2c
    212DE6A4 26251429 RefactoringEssentials_25cb0000!RefactoringEssentials.Util.Analysis.ControlFlowGraphBuilder.BuildControlFlowGraph(Microsoft.CodeAnalysis.CSharp.Syntax.StatementSyntax, Microsoft.CodeAnalysis.SemanticModel, System.Func`3<Microsoft.CodeAnalysis.SyntaxNode,System.Threading.CancellationToken,Microsoft.CodeAnalysis.ISymbol>, System.Threading.CancellationToken)+0xf1
    212DE708 26250901 RefactoringEssentials_25cb0000!RefactoringEssentials.Util.Analysis.ControlFlowGraphBuilder.BuildControlFlowGraph(Microsoft.CodeAnalysis.CSharp.Syntax.StatementSyntax, Microsoft.CodeAnalysis.SemanticModel, System.Threading.CancellationToken)+0x61
    212DE728 26250873 RefactoringEssentials_25cb0000!RefactoringEssentials.Util.Analysis.ReachabilityAnalysis.Create(Microsoft.CodeAnalysis.CSharp.Syntax.StatementSyntax, Microsoft.CodeAnalysis.SemanticModel, RecursiveDetectorVisitor, System.Threading.CancellationToken)+0x23
    212DE740 2624FBDC RefactoringEssentials_25cb0000!RefactoringEssentials.CSharp.Diagnostics.FunctionNeverReturnsAnalyzer+GatherVisitor.VisitBody(System.String, Microsoft.CodeAnalysis.SyntaxToken, Microsoft.CodeAnalysis.CSharp.Syntax.StatementSyntax, Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis.CSharp.SyntaxKind)+0x84
    212DE7BC 2624E227 RefactoringEssentials_25cb0000!RefactoringEssentials.CSharp.Diagnostics.FunctionNeverReturnsAnalyzer+GatherVisitor.VisitMethodDeclaration(Microsoft.CodeAnalysis.CSharp.Syntax.MethodDeclarationSyntax)+0x97
    212DE7E4 264406FD Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.Syntax.MethodDeclarationSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor)+0xd
    212DE7EC 2608AEDC Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker.Visit(Microsoft.CodeAnalysis.SyntaxNode)+0x3c
    212DE7FC 2608C869 Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker.DefaultVisit(Microsoft.CodeAnalysis.SyntaxNode)+0x71
    212DE834 2624CDDB Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor.VisitClassDeclaration(Microsoft.CodeAnalysis.CSharp.Syntax.ClassDeclarationSyntax)+0xb
    212DE83C 263BB7BD Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.Syntax.ClassDeclarationSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor)+0xd
    212DE844 2608AEDC Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker.Visit(Microsoft.CodeAnalysis.SyntaxNode)+0x3c
    212DE854 2608C869 Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker.DefaultVisit(Microsoft.CodeAnalysis.SyntaxNode)+0x71
    212DE88C 263BB79B Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor.VisitNamespaceDeclaration(Microsoft.CodeAnalysis.CSharp.Syntax.NamespaceDeclarationSyntax)+0xb
    212DE894 263BB77D Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.Syntax.NamespaceDeclarationSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor)+0xd
    212DE89C 2608AEDC Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker.Visit(Microsoft.CodeAnalysis.SyntaxNode)+0x3c
    212DE8AC 2608C869 Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker.DefaultVisit(Microsoft.CodeAnalysis.SyntaxNode)+0x71
    212DE8E4 263BAAEB Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor.VisitCompilationUnit(Microsoft.CodeAnalysis.CSharp.Syntax.CompilationUnitSyntax)+0xb
    212DE8EC 263BA9DD Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.Syntax.CompilationUnitSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor)+0xd
    212DE8F4 2608AEDC Microsoft_CodeAnalysis_CSharp_1aa00000!Microsoft.CodeAnalysis.CSharp.CSharpSyntaxWalker.Visit(Microsoft.CodeAnalysis.SyntaxNode)+0x3c
    212DE904 2624CD68 RefactoringEssentials_25cb0000!RefactoringEssentials.CSharp.Diagnostics.FunctionNeverReturnsAnalyzer+<>c__DisplayClass3_0+<<Initialize>b__1>d.MoveNext()+0x1e0
```

We will fix VS to handle this gracefully (we are tracking fixing this with https://github.com/dotnet/roslyn/issues/13120), but the analyzers also need to be fixed. This particular analyzer can be fixed by registering a SemanticModelAction instead of a SyntaxTreeAction.

**Comment by Rpinski** [comment]
@mavasani Thank you for the information!

**Comment by Rpinski** [comment]
So far I haven't found more analyzers registering async callbacks (besides those fixed by 62451c4), therefore I close this issue for now.
